### PR TITLE
Add support for Nuvoton NUC472 NuTiny

### PR DESCRIPTION
**Pull request by ccli8**

### CODE REVIEW

File: mcu/nuvoton/nuc472hi8ae.yaml [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+    uvision:
+        TargetOption:
+            Device:
+                - NUC472HI8AE
+            Vendor:
+                - Nuvoton
+            Cpu:
+                - IRAM(0x20000000,0x2000FFFF) IROM(0,0x7FFFF) CLOCK(50000000) CPUTYPE("Cortex-M4") FPU2
+            FlashDriverDll:
+                -
```
Comment (0xc0170) [inline mcu/nuvoton/nuc472hi8ae.yaml:18]: Why is FlashDriver empty and DeviceId 0 ? Did you test uvision project if it selects  the correct MCU and flashes properly? I can't recall FlashDriverDll importance, if you consider it should not be in the definitions (not required), please create an issue for this.

Did you use progendef script to get create this yaml file?

**Comment by 0xc0170** [comment]
Is this device for uvision in the legacy pack or do I need a new one?

File: mcu/nuvoton/nuc472hi8ae.yaml (hunk 1 above)
Comment (ccli8) [inline mcu/nuvoton/nuc472hi8ae.yaml:18]: Our ICE (Nu-Link) doesn't follow CMSIS-DAP. It has private protocol and needs its own ICE driver for Keil installed (only Windows). So some items are left empty.

I didn't use progendef script to create my yaml file. Is it must?

Comment (0xc0170) [inline mcu/nuvoton/nuc472hi8ae.yaml:18]: No it's not a must, was just wondering if that script did not parse data correctly.

I'll test this locally today

Comment (ccli8) [inline mcu/nuvoton/nuc472hi8ae.yaml:18]: You'll be unable to test it. S/W and H/W are not fully ready yet.

**Comment by 0xc0170** [comment]
I found this device in the software packs, which defines this structure for this device:

```
          <Device>NUC472HI8AE</Device>
          <Vendor>Nuvoton</Vendor>
          <PackID>Nuvoton.NuMicro_DFP.1.0.4</PackID>
          <PackURL>http://www.nuvoton.com/hq/enu/Documents/KEILSoftwarePack</PackURL>
          <Cpu>IROM(0x00000000,0x80000) IRAM(0x20000000,0x10000) CPUTYPE("Cortex-M4") FPU2 CLOCK(12000000)</Cpu>
          <FlashUtilSpec></FlashUtilSpec>
          <StartupFile></StartupFile>
          <FlashDriverDll>UL2CM3(-S0 -C0 -P0 -FD20000000 -FC1000 -FN1 -FF0NUC400_AP_512 -FS00 -FL080000 -FP0($$Device:NUC472HI8AE$Flash\NUC400_AP_512.FLM))</FlashDriverDll>
          <DeviceId>0</DeviceId>
          <RegisterFile>$$Device:NUC472HI8AE$Device\NUC400\Include\NUC472_442.h</RegisterFile>
          <MemoryEnv></MemoryEnv>
```

for progendef , register file is important if new sw pack is used , that's the distinction here. And uvision fills flashdriverdll.

**Comment by 0xc0170** [comment]
@ccli8 : Does uvision work for you with this device - using project generator or yotta debug? What is the uvision version you are using? Is is using legacy device database or new software packs?

The code above I referenced is from uvision 5.14 and latest nuvoton software pack downloaded yesterday. To have a new software pack recognized, use RegisterFile needs to be in yaml file (described here https://github.com/project-generator/project_generator_definitions/tree/dev_v1#uvision).

Sorry for all these questions, just trying to understand why Flash Dll and DeviceID are empty. As this is not usual, and might require some changes in the uvision parser and definitions.

**Comment by 0xc0170** [comment]
@ccli8 bump

**Comment by ccli8** [comment]
My uvision version is 4.74. The yaml file is created by referring existent yaml sample, not from progendef.

Per my test, in my uvision project, FlashDriverDll is empty because I select **Nuvoton Nu-Link Debugger**. My NUC472 NuTiny board doesn't support **CMSIS-DAP Debugger**.

I test with **yotta debug** and it works. But it needs the following preparations:
1. Install **Nu-Link Keiil Driver** at the link below:
http://www.nuvoton.com/hq/products/microcontrollers/arm-cortex-m4-mcus/Software/?__locale=en&resourcePage=Y
2. Nuvoton NUC472 NuTiny mbed porting. This is on-going.
3. Nuvoton NUC472 NuTiny board.
4. Add Nuvoton NUC472 NuTiny support as this Pull Request into project_generator_definitions.
5. Run _yotta debug_. In uvision, select **Nuvoton Nu-Link Debugger** rather than default **CMSIS-DAP Debugger**.

**Comment by 0xc0170** [comment]
Thanks for explanation above. I'll merge this as it is, release a new patch.

> Run yotta debug. In uvision, select Nuvoton Nu-Link Debugger rather than default CMSIS-DAP Debugger.

There is an issue for valinor to pass a debugger to progen. progen supports debugger options, however Nuvoton Nu-Link Debugger is not there. Therefore you can add it (uvision debuggers are defined here https://github.com/project-generator/project_generator/blob/master/project_generator/tools/uvision.py#L35). Once valinor/yotta passes debugger option, yt debug should work out of the box.

**Pull request merged**